.github/workflows/ci.yml

```diff
@@ -325,36 +325,36 @@ jobs:
       - name: "Fetching cache (CMake/ccache)"
         uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
         with:
-          path: ${{ github.workspace }}/.ccache
-          key: ccache_all_macos_${{ github.sha }}
-          restore-keys: ccache_all_macos
+          path: ${{ env.CCACHE_DIR }}
+          key: ${{ env.CCACHE_GH_CACHE_KEY }}_${{ github.sha }}
+          restore-keys: ${{ env.CCACHE_GH_CACHE_KEY }}
       # Finally: build and run tests.
       - name: "Building IREE"
         env:
           IREE_READ_REMOTE_CCACHE: 0
           IREE_WRITE_REMOTE_CCACHE: 0
           IREE_READ_LOCAL_CCACHE: 1
-          IREE_WRITE_LOCAL_CCACHE: ${{ needs.setup.outputs.write-caches }}
-          CCACHE_DIR: ${{ github.workspace }}/.ccache
+          IREE_WRITE_LOCAL_CCACHE: 1 # DO NOT SUBMIT
           # Cache size and compression level settings are a delicate balance.
           # * A full build cache is around 2-5GB depending on compression level
           # * Upload/download is slow (double compression may or may not help)
           # * We have a limit of 10GB across all cached files per repository
-          # * Cache misses are quite costly:
-          #   * 99% cache hits -> ~5 minutes to build
-          #   * 20% cache hits -> ~15-20 minutes to build
+          # * Cache misses are quite costly. Cache restore on Macs is painfully
+          #   slow and grinds to a complete halt somewhere around a cache size
+          #   of 2GB. See https://github.com/actions/cache/issues/1115.
+          #   Hopefully this is still large enough to be somewhat useful.
           CCACHE_MAXSIZE: 4G
-          CCACHE_COMPRESSLEVEL: 5
         run: bash ./build_tools/cmake/build_all.sh "${BUILD_DIR}"
       - name: "Testing IREE"
         run: bash ./build_tools/cmake/ctest_all.sh "${BUILD_DIR}"
       # Write cache (if configured to) after all other steps are finished.
+      - name: "Recompressing cache (CMake/ccache)"
+        run: ccache --recompress 10
       - name: "Saving cache (CMake/ccache)"
-        if: needs.setup.outputs.write-caches == '1'
         uses: actions/cache/save@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
         with:
-          path: ${{ github.workspace }}/.ccache
-          key: ccache_all_macos_${{ github.sha }}
+          path: ${{ env.CCACHE_DIR }}
+          key: ${{ env.CCACHE_GH_CACHE_KEY }}_${{ github.sha }}
```

Comment:
https://github.com/openxla/iree/actions/runs/4623460718/jobs/8178728862?pr=12905#step:11:18

`Failed to save: Unable to reserve cache with key ccache_all_macos_cb7ce722eb10c896195ff203c9be77005c7457d8, another job may be creating this cache. More details: Cache already exists. Scope: refs/pull/12905/merge, Key: ccache_all_macos_cb7ce722eb10c896195ff203c9be77005c7457d8, Version: 9f9bec005f64399df969515b921f6f040690a403ecdee39a26aec50792275dbe`

Reply:
Oh ffs. I guess the cache key has to be unique per *run*? That's so stupid

Reply:
Ah looks like the issue comes from https://github.com/actions/toolkit/issues/658. Seems cache keys really do have to be unique. That scuttles our plans to cache based on LLVM commit. Wow that is so stupid...